### PR TITLE
Reading of Metadata and pdb blobs via Soft debugger protocol

### PR DESCRIPTION
**Pull request by nerzhulart**
This PR allows debugger clients to retrieve metadata and pdbs as blobs via soft debugger protocol. Debugger may use own metadata/symbol reader which makes possible to extract more information than *Mirror objects provide.
Added commands to get TypeMirror and MethodMirror from assembly by metadata token

### CODE REVIEW

**Comment by vargaz** [comment]
Will review.
In general, it is better to obtain these on the client side, since these files are huge, and transferring them over (possibly slow) links is not good for performance.

**Comment by shalupov** [comment]
@vargaz Generally yes, but if don't have them at the client side it turns into usability nightmare

**Comment by nerzhulart** [comment]
@vargaz Generally it's impossible to get metadata, e.g. when you attached to a remote machine or device. So we want to have unified way to read metadata/symbols, because Mirror API is very poor. We've checked the performance - e.g. transferring of mscorlib (which are quite huge - about 4 mb) takes about 10ms on our machines when debugging locally, so I guess it's won't be a bottleneck. Also debugger may read files when they're accessible locally and ask protocol when not

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/AssemblyMirror.cs [hunk 1]
```diff
@@ -128,5 +141,68 @@ public TypeMirror GetType (String name, Boolean throwOnError)
...
+
+				isDynamic = vm.conn.Assembly_IsDynamic(id);
+				return isDynamic.Value;
+			}
+		}	
+		
+		public bool HasPdb {
+			get {			
+				return GetPdbBlob().Length > 0;
+			}
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/AssemblyMirror.cs]: A method named HasSomething probably shouldn't load mb-s of data from the network.

File: mono/mini/debugger-agent.c [hunk 2]
```diff
@@ -394,7 +395,12 @@ typedef enum {
 	CMD_ASSEMBLY_GET_OBJECT = 4,
 	CMD_ASSEMBLY_GET_TYPE = 5,
 	CMD_ASSEMBLY_GET_NAME = 6,
-	CMD_ASSEMBLY_GET_DOMAIN = 7
+	CMD_ASSEMBLY_GET_DOMAIN = 7,
+    CMD_ASSEMBLY_GET_METADATA_BLOB = 8,
+    CMD_ASSEMBLY_GET_IS_DYNAMIC = 9,
```
Comment (vargaz) [inline mono/mini/debugger-agent.c]: space vs tabs problem.

File: mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/AssemblyMirror.cs [hunk 3]
```diff
@@ -15,8 +16,13 @@ public class AssemblyMirror : Mirror
 		AssemblyName aname;
 		AssemblyDefinition meta;
 		AppDomainMirror domain;
+		byte[] metadata_blob;
+		bool? isDynamic;
+		byte[] pdb_blob;
 		Dictionary<string, long> typeCacheIgnoreCase = new Dictionary<string, long> (StringComparer.InvariantCultureIgnoreCase);
 		Dictionary<string, long> typeCache = new Dictionary<string, long> ();
+		Dictionary<uint, long> tokenTypeCache = new Dictionary<uint, long>();
+		Dictionary<uint, long> tokenMethodCache = new Dictionary<uint, long>();
```
Comment (vargaz) [inline mcs/class/Mono.Debugger.Soft/Mono.Debugger.Soft/AssemblyMirror.cs]: Please use the mono coding conventions i.e. space before '(' both in the c# and c code.

**Comment by vargaz** [comment]
Looks ok otherwise.

**Comment by Iliya-usov** [comment]
@alexanderkyte @vargaz Fixed

**Comment by marek-safar** [comment]
approve

**Comment by marek-safar** [comment]
@monojenkins commit apidiff

**Comment by marek-safar** [comment]
Thank you very much for the contribution!

**Comment by vargaz** [comment]
This caused a regression since parts of Mono.Debugger.Soft check Metadata against null. This PR will fix it:
https://github.com/mono/mono/pull/9346

**Pull request merged**